src/js/components/Tabs/Tabs.js

```diff
@@ -29,16 +42,226 @@ const Tabs = forwardRef(
+          headerRef.current.scrollBy({
+            left: amountHidden,
+            behavior: 'smooth',
+          });
+        }
+
+        // wait for scroll animation to finish
+        // checks every 50 milliseconds for 1000 milliseconds
+        // if the scroll animation has finished. Most scroll
+        // animations will finish in 1000 milliseconds unless
+        // the tab name is very long.
+        if (isSafari) {
+          updateArrowState();
+        } else {
+          const checkVisible = setInterval(() => {
+            if (tabRefs[index].current && isVisible(tabRefs[index].current)) {
+              updateArrowState();
+              clearInterval(checkVisible);
+            }
+          }, 50);
+          setTimeout(() => {
+            updateArrowState();
+            clearInterval(checkVisible);
+          }, 1000);
+        }
+      },
+      [tabRefs, headerRef, isVisible, updateArrowState, isSafari],
+    );
+
+    const moveByArrowKey = (direction) => {
+      const previous = direction === 'previous';
+      let index = direction === 'previous' ? 0 : tabRefs.length - 1;
+      let scrolledToIndex;
+      const moveBy = theme.tabs.step[size] - 1 || 0;
+
+      while (
+        scrolledToIndex === undefined &&
+        ((previous && index < tabRefs.length - 1) || (!previous && index > 0))
+      ) {
+        if (
+          !isVisible(tabRefs[index].current) &&
+          ((previous && isVisible(tabRefs[index + 1].current)) ||
+            (!previous && isVisible(tabRefs[index - 1].current)))
+        ) {
+          if (previous) {
+            if (index - moveBy >= 0) {
+              scrollToIndex(index - moveBy, false);
+              scrolledToIndex = index - moveBy;
+            } else {
+              scrollToIndex(0, false);
+              scrolledToIndex = 0;
+            }
+          } else if (index + moveBy < tabRefs.length) {
+            scrollToIndex(index + moveBy, false);
+            scrolledToIndex = index + moveBy;
+          } else {
+            scrollToIndex(tabRefs.length - 1, false);
+            scrolledToIndex = tabRefs.length - 1;
+          }
+        }
+        index = previous ? index + 1 : index - 1;
+      }
+    };
+
+    useEffect(() => {
+      // if the active tab isn't visible scroll to it
+      if (
+        overflow &&
+        tabRefs &&
+        tabRefs[activeIndex].current &&
+        !isVisible(tabRefs[activeIndex].current)
+      )
+        scrollToIndex(activeIndex, true);
+      // eslint-disable-next-line react-hooks/exhaustive-deps
+    }, [overflow, activeIndex]);
+
+    useEffect(() => {
+      // if the focus index is changing make sure the previous and next arrows
+      // are in sync
+      if (
+        overflow &&
+        focusIndex !== -1 &&
+        !isVisible(tabRefs[focusIndex].current)
+      )
+        scrollToIndex(focusIndex, true);
+      else if (overflow && focusIndex !== -1) {
+        // If the browser scrolled the focused item into view and
```

Comment:
Do we still need this block? Since I would assume `scrollToIndex` and `isVisible` would already be taking care of this situation.

Reply:
This block is used in case where the browser handles the scroll into view of focused items. In this case `scrollTo` won't be called because the tab will be visible so we need to handle scrolling a few pixels further to show the focusIndicator

Reply:
I would expect `isVisible()` to take care of that already, such that we would call `scrollTo` at line 206 if any scrolling might be needed. In other words, I would expect `isVisible` of true to mean we don't need to scroll and `isVisible` of false meaning we do need to scroll.

Reply:
Maybe we should rename `isVisible` to `needsScroll`?